dash/development/_py_prop_typing.py

```diff
@@ -0,0 +1,170 @@
+import collections
+import json
+import string
+import textwrap
+
+import stringcase
+
+
+shapes = {}
+shape_template = """{name} = TypedDict(
+    "{name}",
+    {values}
+)
+"""
+custom_imports = collections.defaultdict(lambda: collections.defaultdict(list))
+
+
+def _clean_key(key):
+    k = ""
+    for ch in key:
+        if ch not in string.ascii_letters + "_":
+            k += "_"
+        else:
+            k += ch
+    return k
+
+
+def generate_any(*_):
+    return "typing.Any"
+
+
+def generate_shape(type_info, component_name: str, prop_name: str):
+    props = []
+    name = stringcase.pascalcase(prop_name)
+
+    for prop_key, prop_type in type_info["value"].items():
+        typed = get_prop_typing(
+            prop_type["name"], component_name, f"{prop_name}_{prop_key}", prop_type
+        )
+        if not prop_type.get("required"):
+            props.append(f'        "{prop_key}": NotRequired[{typed}]')
+        else:
+            props.append(f'        "{prop_key}": {typed}')
+
+    shapes.setdefault(component_name, {})
+    shapes[component_name][name] = textwrap.indent(
+        shape_template.format(
+            name=name, values="    {\n" + ",\n".join(props) + "\n    }"
+        ),
+        "    ",
+    )
+
+    return f'"{name}"'
+
+
+def generate_union(type_info, component_name: str, prop_name: str):
+    types = []
+    for union in type_info["value"]:
+        u_type = get_prop_typing(union["name"], component_name, prop_name, union)
+        if u_type not in types:
+            types.append(u_type)
+    return f"typing.Union[{', '.join(types)}]"
+
+
+def generate_tuple(
+    type_info,
+    component_name: str,
+    prop_name: str,
+):
+    els = type_info.get("elements")
+    elements = ", ".join(
+        get_prop_typing(x.get("name"), component_name, prop_name, x) for x in els
+    )
+    return f"typing.Tuple[{elements}]"
+
+
+def generate_array_of(
+    type_info,
+    component_name: str,
+    prop_name: str,
+):
+    typed = get_prop_typing(
+        type_info["value"]["name"], component_name, prop_name, type_info["value"]
+    )
+    return f"typing.Union[typing.Sequence[{typed}], typing.Tuple]"
```

Comment:
Why do we need the `Union` here? `Sequence` includes tuples, I believe?

Reply:
Good catch, I changed from List but didn't remove the Tuple.